halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/AccountService.java

```diff
@@ -37,8 +39,13 @@
   @Autowired
   ProviderService providerService;
 
-  public Account getAccount(String deploymentName, String providerName, String accountName) {
-    Provider provider = providerService.getProvider(deploymentName, providerName);
+  @Autowired
+  ValidateService validateService;
+
+  public Account getAccount(NodeReference reference) {
+    Provider provider = providerService.getProvider(reference);
```

Comment:
I'm still finding this API weird. I'd expect reference.getProvider().
I guess this ok because either way you are getting the provider. But since you only autowire the provider service to give it the reference to get the provider, it makes this that much more awkward. A reference sounds like a data object so this getProvider() wouldnt be appropriate there. But if this is a common pattern for resolving objects in the reference, then NodeReference could have a companion class that you use instead and that would autowire all the things instead. I dont know if that's a good idea or not. Just a reaction reading this.
The difference again is one of encapsulating the implementation details. The interface is the clique "NodeReference" and "NodeReferenceResolver" (bad name) rather than NodeReference and *Service.

On the other hand, at least in this case, AccountService and ProviderService are already coupled since Accounts are part of Providers. So in this case it isnt as big a deal, though still feels like something is going on here and, at least for now, you dont actually use the service for anything other than resolving the reference.

Leave it as is for now and see how it bakes out. I am just sharing thoughts and gut reaction for now.